### PR TITLE
perf: do not deep clone upfront

### PR DESCRIPTION
**Pull request by P0lip**
cloneDeep is fairly expensive in certain cases and does not need to be implement in the way it is now, since we visit each object, and hence can simply shallow copy it before modyfing.
Using shallow copy of each object before visiting it significantly (by around 25-35%) improves the perfomance of `@stoplight/http-spec#transformOas3Operations`. Tested against various specs, such as Stripe, Whatsapp, etc.
Obviously, the results will vary - if a given document does not have schemas or has a handful of small schemas, the difference is likely to be less notable.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,6 +1,5 @@
 var deepEqual = require('fast-deep-equal');
 var convert = require('./lib/convert');
-var clone = require('lodash.clonedeep');
```
Comment (P0lip) [inline index.js:3]: oh, and we no longer need this dependency

File: test/clone_schema.test.js [hunk 2]
```diff
@@ -32,6 +56,8 @@ test('cloning schema with cloneSchema option', function (assert) {
     nullable: true
   }
 
+  var cloned = JSON.parse(JSON.stringify(schema))
```
Comment (philsturgeon) [inline test/clone_schema.test.js:59]: This looks like it’s reversing #9 which @XVincentX did to help circular references. Can we make sure we’re not speeding things up by breaking some edge cases? 😅

Comment (XVincentX) [reply test/clone_schema.test.js:59]: The general problem is that this library is mutating the passed object; the `cloneSchema` thing happens exclusively in case you want immutability from the outside so that the original object stays the same. It's a band-aid solution.

The best solution here would be to simply stop mutating the object in place using the fp counterpart of lodash (such as https://gist.github.com/brauliodiez/9fd567fdc8b2695c11a61daa50475f43), but it would require a significant change in this library (and likely a breaking change as well)

File: lib/converters/schema.js [hunk 3]
```diff
@@ -13,7 +13,11 @@ function convertFromSchema (schema, options) {
 }
 
 function convertSchema (schema, options) {
-	var structs = options._structs;
+  if (options.cloneSchema) {
```
Comment (XVincentX) [inline lib/converters/schema.js:16]: How come did you decide to move this check here?

File: lib/converters/schema.js [hunk 4]
```diff
@@ -92,17 +103,14 @@ function convertProperties (properties, options) {
...
 
 		if (!isObject(property)) {
 			continue
 		}
 
-		options._removeProps.forEach(function (prop) {
-			if (property[prop] === true) {
-				removeProp = true
-			}
+    removeProp = options._removeProps.some(function (prop) {
```
Comment (XVincentX) [inline lib/converters/schema.js:112]: 👍

**Comment by P0lip** [comment]
@philsturgeon @XVincentX any of you eager to review this one?

**Comment by P0lip** [comment]
> Can you do one example of a clone deep that is not required for our usages? I can't really think of one at the moment

Technically we're doing deep cloning here - we copy each object we visit before mutating it.
The PR title as well as the description might be a bit slightly misleading.
What we're doing here is performing a shallow copy of the object we visit.
This has a number of benefits:
- we do not iterate over the whole, often huge, object twice
- we do not clone the values that won't be included in the final object
- no extra dep :P

> If the sole aim is to avoid clone deep, maybe we can use lodash.clone? That'd make me feel more comfortable instead of our own implementation.

I'm strongly against using any of lodash clones in this particular case. They are great and take care of different object types such as Date, Maps, Sets, etc., but we do not need to clone any of them.

**Review by XVincentX — APPROVED** [review]

**Comment by P0lip** [comment]
@philsturgeon could you merge this one assuming you are fine with the changes?

**Review by philsturgeon — APPROVED** [review]
